### PR TITLE
WIP: remove default formfield choices form formfield model, add formfielda…

### PR DESCRIPTION
**Pull request by benzkji**
broken tests, but as a base for discussion. `FormField.type` will not validate, as the only valid choice is the empty string.

...also, changed db to local file, seems that test/tox stil uses an in memory one.

### CODE REVIEW

**Comment by benzkji** [comment]
#30

File: form_designer/admin.py [hunk 1]
```diff
@@ -14,6 +14,11 @@
 from form_designer import models
 from xlsxdocument import XLSXDocument
 
+from .models import FIELD_TYPES
+
+
+FIELD_TYPE_CHOICES = ((type["type"], type["verbose_name"]) for type in FIELD_TYPES)
```
Comment (matthiask) [inline form_designer/admin.py:20]:
```suggestion
FIELD_TYPE_CHOICES = ((type["type"], type["verbose_name"]) for type in models.FIELD_TYPES)
```

models are already available in the file, no need to import them again.

File: tests/testapp/settings.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import os
 
-DATABASES = {"default": {"ENGINE": "django.db.backends.sqlite3", "NAME": ":memory:"}}
+DATABASES = {"default": {"ENGINE": "django.db.backends.sqlite3", "NAME": "dev.db"}}
```
Comment (matthiask) [inline tests/testapp/settings.py:3]: Not sure why you want to change this?

File: tests/testapp/settings.py [hunk 3]
```diff
@@ -15,6 +15,7 @@
     "form_designer",
 ]
 
+DEBUG = True
```
Comment (matthiask) [inline tests/testapp/settings.py:18]: Let's continue running tests with `DEBUG=False` please.

File: tests/testapp/test_forms.py [hunk 4]
```diff
@@ -186,6 +186,7 @@ def test_admin(self):
 
         data["fields-0-choices"] = ""
         response = self.client.post("/admin/form_designer/form/add/", data)
+        print(response.content)
```
Comment (matthiask) [inline tests/testapp/test_forms.py:189]:
```suggestion
```

**Review by matthiask — COMMENTED** [review]
I'm really happy when this is all it took! I added a few comments to the pull request.

File: tests/testapp/settings.py (hunk 2 above)
Comment (benzkji) [reply tests/testapp/settings.py:3]: to be able to have a local test database, so I can do a manage.py runserver, and actually see things, not only test them ;-) it looks like tox and manage.py test still use an in memory db? as an alternative, we could use a separate settings.py for local dev, but if it just works, I would prefer this one? up to you.

File: tests/testapp/settings.py (hunk 3 above)
Comment (benzkji) [reply tests/testapp/settings.py:18]: sorry, this is another side effect of my "run it locally" mode...so it's probably better to have a separate settings_dev.py

File: tests/testapp/test_forms.py (hunk 4 above)
Comment (benzkji) [reply tests/testapp/test_forms.py:189]: outch.

**Comment by benzkji** [comment]
the main point is: it will probably be needed to remove the `choices=(("", ""), )` from the models type field. is the `get_field_display()` really needed somewhere?

File: tests/testapp/test_forms.py (hunk 4 above)
Comment (benzkji) [reply tests/testapp/test_forms.py:189]: this is where it breaks ;-)

**Comment by benzkji** [comment]
it will still fail in CI, as tox fetches django 3.1, where the tests/form_designer are/is broken. they work with django 2.2/3.0.

**Comment by matthiask** [comment]
> it will still fail in CI, as tox fetches django 3.1, where the tests/form_designer are/is broken. they work with django 2.2/3.0.

Yeah, it fails with Django@master. However, that's because of FeinCMS – form designer works well with Django 3.1 and master.

**Comment by benzkji** [comment]
aahhh. thought about proposing removing feincms from tests, but that's another story. what do you think, could we remove the choices from `type`? I even tried AppConfig.ready(), but am afraid of messing with _meta, and attributes where my debugger says "dont touch me".

**Comment by benzkji** [comment]
and, settings. I myself prefer the convenience of a ./manage.py runserver that just works. if we change the db, and it still works in tests, the only needed thing is ALLOWED_HOSTS with localhost, and then we can keep DEBUG=False. and provide a settings_debug.py, when we really need it...?

**Pull request closed without merging**